File: ProtocolAnalysis/src/ProtocolAnalysis/EventHandle/EventDispatch.py
#-*- encoding=utf-8 -*


from ProtocolAnalysis.DelayHandle.DelayHandleMgrBase import DelayHandleMgrBase
from ProtocolAnalysis.DataStruct.MList import MList
from ProtocolAnalysis.EventHandle.EventDispatchFunctionObject import EventDispatchFunctionObject
from ProtocolAnalysis.Core.AppSysBase import AppSysBase


'''
@brief 事件分发，之分发一类事件，不同类型的事件使用不同的事件分发
@brief 注意，事件分发缺点就是，可能被调用的对象已经释放，但是没有清掉事件处理器，结果造成空指针
'''
class EventDispatch(DelayHandleMgrBase):
    '''
    classdocs
    '''


    def __init__(self, eventId_ = 0):
        '''
        Constructor
        '''
        super(EventDispatch, self).__init__()
        
        self.m_eventId = eventId_
        self.m_handleList = MList()
        self.m_uniqueId = 0       # 唯一 Id ，调试使用


    # 相同的函数只能增加一次
    def addEventHandle(self, handle, paramList):
        funcObject = EventDispatchFunctionObject();
        funcObject.m_handle = handle;
        funcObject.m_paramList = paramList
        if None != handle:
            self.addObject(funcObject);
        else:
            AppSysBase.instance().m_logSys.log("Event Handle is null");


    def addObject(self, delayObject, priority = 0.0):
        if self.bInDepth():
            super(EventDispatch, self).addObject(delayObject, priority);
        else:
            # 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
            self.m_handleList.Add(delayObject);


    def removeEventHandle(self, handle):
        for idx in range(0, self.m_handleList.Count(), 1):
            if self.m_handleList[idx].m_handle is handle:
                break
        if idx < self.m_handleList.Count():
            self.delObject(self.m_handleList[idx])
        else:
            AppSysBase.instance().m_logSys.log("Event Handle not exist");


    def delObject(self, delayObject):
        if self.bInDepth():
            super(EventDispatch, self).delObject(delayObject);
        else:
            if not self.m_handleList.Remove(delayObject):
                AppSysBase.instance().m_logSys.log("Event Handle not exist");


    def dispatchEvent(self, dispatchObject):
        self.incDepth();

        for handle in self.m_handleList.list:
            if not handle.m_bClientDispose:
                handle.m_handle(handle.m_paramList, dispatchObject);

        self.decDepth();


    def clearEventHandle(self):
        if self.bInDepth():
            for item in self.m_handleList.list:
                self.delObject(item)
        else:
            self.m_handleList.Clear()


    # 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
    def existEventHandle(self, handle):
        bFinded = False
        for item in self.m_handleList.list:
            if item.m_handle is handle:
                bFinded = True
                break;

        return bFinded


    def copyFrom(self, rhv):
        for handle in rhv.handleList.list:
            self.m_handleList.Add(handle)
 

File: LuaLibs/src/MyLua/Libs/EventHandle/EventDispatch.lua
--[[
    @brief 事件分发器
]]

require "MyLua.Libs.Core.GlobalNS"
require "MyLua.Libs.Core.Class"
require "MyLua.Libs.DataStruct.MList"
require "MyLua.Libs.EventHandle.EventDispatchFunctionObject"

local M = GlobalNS.Class(GlobalNS.DelayHandleMgrBase)
M.clsName = "EventDispatch"
GlobalNS[M.clsName] = M

function M:ctor(eventId_)
    self.m_eventId = eventId_
    self.m_handleList = GlobalNS.MList:new()
    self.m_uniqueId = 0       -- 唯一 Id ，调试使用
end

function M:dtor()

end

function M:getHandleList()
    return self.m_handleList;
end

function M:getUniqueId()
    return self.m_uniqueId;
end

function M:setUniqueId(value)
    self.m_uniqueId = value;
    --self.m_handleList.uniqueId = m_uniqueId;
end

function M:addEventHandle(handle, pThis)
    local funcObject = GlobalNS.EventDispatchFunctionObject:new()
    funcObject.m_handle = handle;
    funcObject.m_pThis = pThis;
    if (nil ~= handle) then
        self:addObject(funcObject);
    else
        -- 日志
    end
end

function M:addObject(delayObject, priority)
    if self:bInDepth() then
        self.super.addObject(self, delayObject, priority); -- super 使用需要自己填充 Self 参数
    else
        -- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
        self.m_handleList:Add(delayObject);
    end
end

function M:removeEventHandle(handle, pThis)
    local idx = 0;
    for idx = 0, idx < self.m_handleList:Count(), 1 do
        if self.m_handleList:at(idx).m_handle == handle and self.m_handleList.at(idx).m_pThis == pThis then
            break;
        end
    end
    if idx < self.m_handleList:Count() then
        self:delObject(self.m_handleList[idx]);
    else
        -- 日志
    end
end

function M:delObject(delayObject)
    if self:bInDepth() then
        self.super.delObject(self, delayObject);
    else
        if self.m_handleList:Remove(delayObject) == false then
            -- 日志
        end
    end
end

function M:dispatchEvent(dispatchObject)
    self:incDepth();

    for _, handle in ipairs(self.m_handleList:list()) do
        if handle.m_bClientDispose == false then
            handle:call(dispatchObject);
        end
    end

    self:decDepth();
end

function M:clearEventHandle()
    if (self:bInDepth()) then
        for _, item in ipairs(self.m_handleList:list()) do
            self:delObject(item);
        end
    else
        self.m_handleList:Clear();
    end
end

-- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
function M:existEventHandle(handle)
    local bFinded = false;
    for _, item in ipairs(self.m_handleList:list()) do
        if (item.m_handle == handle) then
            bFinded = true;
            break;
        end
    end

    return bFinded;
end

function M:copyFrom(rhv)
    for _, handle in ipairs(rhv.handleList:list()) do
        self.m_handleList:Add(handle);
    end
end

return M